akka-docs/rst/java/code/jdocs/akka/typed/IntroSpec.java

```diff
@@ -0,0 +1,229 @@
+          sessions.forEach(s -> s.tell(mp));
+          return Actor.same();
+        } else {
+          return Actor.unhandled();
+        }
+      });
+    }
+    //#chatroom-behavior
+
+    public static Behavior<Command> mutableChatRoom() {
+      return Actor.deferred(c -> {
+        final List<ActorRef<SessionEvent>> sessions = new ArrayList<ActorRef<SessionEvent>>();
+        int postCount = 0;
+
+        return Actor.stateful((ctx, msg) -> {
+          if (msg instanceof GetSession) {
+            GetSession getSession = (GetSession) msg;
+            ActorRef<PostMessage> wrapper = ctx.createAdapter(p ->
+              new PostSessionMessage(getSession.screenName, p.message));
+            getSession.replyTo.tell(new SessionGranted(wrapper));
+            sessions.add(getSession.replyTo);
+            return Actor.same();
+          } else if (msg instanceof PostSessionMessage) {
+            PostSessionMessage post = (PostSessionMessage) msg;
+            MessagePosted mp = new MessagePosted(post.screenName, post.message);
+            // Local variable postCount defined in an enclosing scope must be final or effectively final
+            //postCount += 1;
+            sessions.forEach(s -> s.tell(mp));
+            return Actor.same();
+          } else {
+            return Actor.unhandled();
+          }
+        });
+      });
+    }
+
+    public static Behavior<Command> mutableChatRoom2() {
+      // TODO Type safety: The expression of type IntroSpec.ChatRoom.MutableChatRoom2 needs unchecked conversion to
+      //      conform to Behavior<IntroSpec.ChatRoom.Command>
+      return Actor.deferred(ctx -> new MutableChatRoom2());
+    }
+
+    public static class MutableChatRoom2 extends ExtensibleBehavior<Command> {
+      final List<ActorRef<SessionEvent>> sessions = new ArrayList<ActorRef<SessionEvent>>();
+      int postCount = 0;
+
+      @Override
+      public Behavior<Command> message(akka.typed.ActorContext<Command> ctx, Command msg) {
+        if (msg instanceof GetSession) {
+          GetSession getSession = (GetSession) msg;
+          ActorRef<PostMessage> wrapper = ctx.createAdapter(p ->
+            new PostSessionMessage(getSession.screenName, p.message));
+          getSession.replyTo.tell(new SessionGranted(wrapper));
+          sessions.add(getSession.replyTo);
+          return Actor.same();
+        } else if (msg instanceof PostSessionMessage) {
+          PostSessionMessage post = (PostSessionMessage) msg;
+          MessagePosted mp = new MessagePosted(post.screenName, post.message);
+          postCount += 1;
+          sessions.forEach(s -> s.tell(mp));
+          return Actor.same();
+        } else {
+          return Actor.unhandled();
+        }
+      }
+
+      @Override
+      public Behavior<Command> management(akka.typed.ActorContext<Command> ctx, Signal msg) {
+        return Actor.same();
+      }
+    }
+
+    public static Behavior<Command> mutableChatRoom3() {
```

Comment:
I like this a lot. Several up-sides from what I can see. Perhaps some of them are too trivial to even mention, but let's do so anyway:

- looks much like old Actors, but now has types. Very likely many users will migrate to this first, and we avoid the risk of being too super different with the new API scaring away people (both new and old). Seems like refactoring many untyped actors to this should be pretty simple, good.
- AFAICS it seems such mutable behaviour could be trivially reused in a fully functional otherwise Actor.
- Very easy to become a mutable behaviour from a stateful/stateless one: `return Actor.stateful((ctx, command) -> { /*...*/ return new MutableChatRoom3(ctx); });` so +1 for reusability

Also, see this draft which makes it possible to write `Actor.mutable(ctx -> msg -> Actor.same())` (thought then it would be more like Actor.create...): https://github.com/ktoso/akka/commit/bd2e89408823c3f3bf3e8f15045833d6075503d5#commitcomment-21797530 This is more for inspiration how these things can be pulled off with target types rather than direclty for the PR.